### DISCUSSION TITLE
Add tab for edit/view the realm "Attributes" in the Admin Console v2

### DISCUSSION THREAD

**Discussion by mpalavrov** [keycloak/keycloak#23678] [Archive]
I would like to suggest to add a tab for viewing/editing the Realm **Attributes** in the Admin Console v2, similar to what we have in our custom Admin Console with the Admin Console v2:
![image](https://github.com/keycloak/keycloak/assets/33352511/746f9eff-82c6-4d1a-ad8f-4f1de15ad428)

**Comment by danielFesenmeyer** [comment]
Related discussion: https://github.com/keycloak/keycloak/discussions/23486

**Reply by danielFesenmeyer** [reply]
And related issue: https://github.com/keycloak/keycloak/issues/17732

**Comment by ssilvert** [comment]
@mpalavrov @danielFesenmeyer I like the idea and it should be quick and easy to implement if one of you wants to submit a PR to implement #17732.

The only thing that I wonder is if there was some reason for not implementing it years ago.  @stianst @mposolda ?

If we do this I'd like to avoid creating a new main tab off of Realm Settings.  We have too many tabs there already.  I suggest breaking Realm settings -> General into two subtabs.  The two subtabs would be "Settings" and "Attributes".  The "Settings" subtab would just be what we have now for General.  And then the "Attributes" subtab would be where you edit the realm attributes.  

We already have a React component for editing attributes.  See use of KeyValueInput on the General tab for an example.

**Comment by stianst** [comment]
Realm attributes are used behind the scenes for a lot of configuration options exposed elsewhere. This can be seen in the screenshot in the description. All these attributes are just exposing the configuration option exposed elsewhere. To 99.99% of users having realm attributes exposed in the UI is just confusing. 

There's a few folks that use it to configure custom extensions with their own custom attributes, but when we've discussed this in the past the use-cases have actually been to configure other type of providers, where it would make more sense to do the config at the provider level. I'd rather improve on support for configuring providers instead of exposing adding realm attributes, which honestly I see as a bit of an anti-pattern.

**Reply by danielFesenmeyer** [reply]
I understand that most config can (and should) be done with providers. However, there's just some configuration shared between providers, which is therefore not specific to a certain provider. That's where we currently use realm attributes. But probably we could just define those shared config on some kind of shared, more generic provider.

**Reply by stianst** [reply]
I'm not totally against some form of realm wide config that can be shared across multiple providers. However, I don't like the idea of just exposing all realm  attributes. If we can come up with something that can register attributes that should be available (should be allow list kinda thing, rather than exclude listing), and ideally some way for providers themselves to register attributes I would be open to adding something.

**Comment by Bredstone** [comment]
Well, I had this same problem a while ago. Me and my team were using the realm attributes to configure some custom requiredAction providers and we wanted them to be accessible through the admin console. I did ask if this would be a good idea and @stianst told me the same as he did now :) 

The realm attributes are used for several realm configs that should not be exposed to users, even admins.

So, in order to solve this, I made a custom admin console theme that adds a new tab to realm settings page. In this tab, only the realm attributes added by the admin are listed. I did this by simply adding a prefix to all the settings that should be accessible on this page (in my case, I chose `USER.DEFINED.`). So, the theme filters all the realm attributes, select the ones that have this prefix, and list them without the prefix being shown. When a realm attribute is added on the tab, the prefix is added automatically and the user doesn't even have to bother with it.

The counter part is the need to add the prefix in code, to access this attributes, but I think it works...

**Comment by adrian-moisa** [comment]
Any update on this one?

**Comment by ramin-n** [comment]
Hi guys
Any news ?
